plasmapy/tests/helpers/__init__.py

```diff
@@ -0,0 +1,35 @@
+"""Test helper functionality for PlasmaPy and affiliated packages."""
+
+# This file contains several commented out import statements.  These
+# statements will be uncommented out over the course of several pull
+# requests that were each originally part of #728.  The blank lines
+# between the import statements will hopefully simplify automatic merging.
+
+# from plasmapy.tests.helpers.actual import ActualTestOutcome
+
+# from plasmapy.tests.helpers.cases import AttrTestCase, FunctionTestCase, MethodTestCase
+
+from plasmapy.tests.helpers.exceptions import (
+    ExceptionMismatch,
+    InvalidTest,
+    MissingException,
+    MissingWarning,
+    TestFailure,
+    TypeMismatch,
+    UnexpectedException,
+    UnexpectedResult,
+    UnexpectedWarning,
+    WarningMismatch,
+)
```

Comment:
I feel this naming convention is a bit vague.  Why not do something like I suggested in https://github.com/PlasmaPy/PlasmaPy/pull/919#discussion_r504986757.  I would even consider renaming the module to `plasmapy.test.helpers.failures`.

> Maybe instead of calling these `<something>Error` we should call them `<something>Failure`. Since this is really a failure opposed to an error. For reference, `pytest.fail` does...
> 
> ```python
> >>> pytest.fail()
> Failed: <Failed instance>
> ```

----

Additionally, if we're continuing with this explicit failure route, then we might want to create a `plasmapy.fail()` version of [`pytest.fail()`](https://docs.pytest.org/en/stable/_modules/_pytest/outcomes.html#fail), which would allow us fail with our custom failures.

Reply:
I'm hesitant to add `Failure` to the end of the exception names because it should be clear from context that it's a test failure.  Hence, I don't think `Failure` would add any additional information to distinguish between the different modes of test failures.  I plan on getting a lot of test failure messages in the future :joy_cat: and I think the shorter names will help us distinguish between the causes of test failures a little bit more quickly, especially once a developer gets used to them.  Ideally, the test failure name should get to the root of the problem as quickly as possible without unnecessary information.

Another reason is that the word `Failure` can be just a little bit discouraging...at least that's how it affected me when I was first starting out.  

> I would even consider renaming the module to `plasmapy.test.helpers.failures`.

I like this idea!  That is a really good fit except for `InvalidTest` which is more of an error than a test in pytestspeak, but I don't think it matters too much.  The path also shouldn't show up to users during actual test failures, so I don't think it'd have the slight psychological discouragement I mentioned above.

> Additionally, if we're continuing with this explicit failure route, then we might want to create a `plasmapy.fail()` version of [`pytest.fail()`](https://docs.pytest.org/en/stable/_modules/_pytest/outcomes.html#fail), which would allow us fail with our custom failures. ~I think this would be more appropriate than raising custom `Exceptions`.~ `pytest.fail()` is just functionality wrapped around a `raises Failed`.

I'm trying to remember if that's what I did way back in #728!  I don't think I ended up doing this, though.  This is outside the scope of this PR, but I'll make a comment in one of the subsequent pull requests so we don't forget.

Reply:
> I'm hesitant to add `Failure` to the end of the exception names because it should be clear from context that it's a test failure. Hence, I don't think `Failure` would add any additional information to distinguish between the different modes of test failures. I plan on getting a lot of test failure messages in the future 😹 and I think the shorter names will help us distinguish between the causes of test failures a little bit more quickly, especially once a developer gets used to them. Ideally, the test failure name should get to the root of the problem as quickly as possible without unnecessary information.

I disagree with this.  **This goes against our whole concept of explicit is better than implicit.**  All exceptions, except `Exception`, end with `Error` for this reason, likewise for `Warnings`.  Now, `Failure` might not be the best extension, maybe `Failed` since that matches better with `pytest.Failed`...or `Fail`.  When outside the context of `plasmapy.test.helpers.exceptions`, the current variable names would have no context.

Having `Error` at the end of an exception does not hinder its readability, so I doubt having something like `Fail` would hinder its readability.

> Another reason is that the word Failure can be just a little bit discouraging...at least that's how it affected me when I was first starting out.

This is something we all have to get use to when we start testing.  However, it should not supersede the need for explicitly over implicity.  It's on us to encourage new contributors when they're getting test failures, not the code.

Reply:
I renamed the test failure exceptions to have `Fail` at the end, except for `InvalidTestError` which is different than a normal test failure.  I decided against renaming `exceptions.py` to `failures.py` since in pretty much every other package we use `exceptions.py`, and it's less needed since we have `Fail` at the end of the exception names now.